include/deal.II/multigrid/mg_constrained_dofs.h

```diff
@@ -146,13 +169,64 @@ MGConstrainedDoFs::initialize(const DoFHandler<dim,spacedim> &dof)
 {
   boundary_indices.clear();
   refinement_edge_indices.clear();
+  level_constraints.clear();
 
   const unsigned int nlevels = dof.get_triangulation().n_global_levels();
 
-  //At this point refinement_edge_indices is empty.
+  //At this point level_constraint and refinement_edge_indices are empty.
+  level_constraints.resize(nlevels);
   refinement_edge_indices.resize(nlevels);
   for (unsigned int l=0; l<nlevels; ++l)
-    refinement_edge_indices[l] = IndexSet(dof.n_dofs(l));
+    {
+      IndexSet relevant_dofs;
+      DoFTools::extract_locally_relevant_level_dofs(dof,l,relevant_dofs);
+      level_constraints[l].reinit(relevant_dofs);
+
+      // Loop through relevant cells and faces finding those which are periodic neighbors.
+      typename DoFHandler<dim,spacedim>::cell_iterator
+      cell = dof.begin(l),
+      endc = dof.end(l);
+      for (; cell!=endc; ++cell)
+        {
+          if (cell->level_subdomain_id() == numbers::artificial_subdomain_id)
+            continue;
```

Comment:
I find this sort of code easier to read if you write it as 
```
  for (...)
    if (cell->level_subdomain_id() != ...)
      {
         ...
      }
```
Having a `continue` inside the loop body is always a non-local control-flow statement -- it's in essence a `goto` which the CS community has not liked for a good long while for good reasons. I think it's not as bad here because it's right after the beginning of the loop, but it's nevertheless easily avoided.